lib/src/model/puzzle/puzzle_history_storage.dart

```diff
@@ -0,0 +1,229 @@
+  }) async {
+    final list = await _db.query(
+      _dbHistory,
+      where: '''
+      userId = ?
+      ''',
+      whereArgs: [
+        userId?.value ?? _anonUserKey,
+      ],
+    );
+
+    final puzzleList = await _db.query(
+      _dbPuzzle,
+    );
+
+    final puzzles = puzzleList.map((e) {
+      final raw = e['data'] as String?;
+      final json = jsonDecode(raw!);
+      if (json is! Map<String, dynamic>) {
+        throw const FormatException(
+          '[PuzzleHistoryStorage] cannot fetch puzzles from $_dbPuzzle: expected an object',
+        );
+      }
+      return Puzzle.fromJson(json);
+    });
+
+    if (puzzles.isEmpty) return null;
+    final history = list.map(
+      (entry) {
+        final raw = entry['data'] as String?;
+        final angle = entry['angle'] as String?;
+        final date = entry['solvedDate'] as String?;
+        if (raw == null || angle == null || date == null) {
+          throw const FormatException(
+            'PuzzleHistoryStorage: connot fetch puzzles: expected an object',
+          );
+        }
+        final json = jsonDecode(raw);
+        if (json is! Map<String, dynamic>) {
+          throw const FormatException(
+            '[PuzzleHistoryStorage] cannot fetch puzzles: expected an object',
+          );
+        }
+        final data = PuzzleHistoryData.fromJson(json);
+        final puzzleAndResult = data.puzzles.map((entry) {
+          return PuzzleAndResult(
+            puzzle: puzzles.firstWhere(
+              (p) => p.puzzle.id == entry.puzzleId,
+              orElse: () => throw FormatException(
+                '[PuzzleHistoryStorage] Cannot find puzzle ${entry.puzzleId}',
+              ),
+            ),
+            result: entry.result,
+          );
+        }).toIList();
+        return PuzzleHistory(
+          puzzles: puzzleAndResult,
+          angle: puzzleThemeNameMap[angle]!,
+          date: date,
+        );
+      },
+    ).toIList();
+
+    return history.isEmpty ? null : history;
+  }
+
+  Future<void> save({
+    required UserId? userId,
+    required DateTime date,
+    required PuzzleTheme angle,
+    required IList<PuzzleIdAndResult> data,
+    required Puzzle puzzle,
+  }) async {
+    await _db.insert(
+      _dbHistory,
+      {
+        'userId': userId?.value ?? _anonUserKey,
+        'angle': angle.name,
+        'solvedDate': DateFormat('yyyy-MM-dd').format(date),
+        'data': jsonEncode(
+          PuzzleHistoryData(puzzles: data.toISet().toIList()).toJson(),
+        ),
+      },
+      conflictAlgorithm: ConflictAlgorithm.replace,
+    );
+    await _db.insert(
+      _dbPuzzle,
+      {
+        'puzzleId': puzzle.puzzle.id.toString(),
+        'lastModified': DateFormat('yyyy-MM-dd').format(date),
+        'data': jsonEncode(puzzle.toJson()),
+      },
+      conflictAlgorithm: ConflictAlgorithm.replace,
+    );
+  }
+}
+
+@Freezed(fromJson: true, toJson: true)
+class PuzzleHistory with _$PuzzleHistory {
```

Comment:
If I understand well your code, this class represents a single day of puzzle history. So the name is not good, it should be called `PuzzleHistoryDay` or sth like that.
And the `date` field should be called `day`.

I'm not sure this is the best way to represent the history actually, but I think to think about it.